lib/tree.rb

```diff
@@ -12,44 +12,130 @@ def initialize(key, val)
+    return nil if @root.nil?
+    #Return the root value if it is the value we are looking for
+    return @root.value if @root.key == key
+    #Otherwise, call in reinforcements
+    return find_helper(@root, key)
+  end
+  
+  def find_helper(current, key)
+    # If we are already on nil, just add it here
+    return nil if current.nil?
+    #if our current is now the same as the key, return it
+    return current.value if current.key == key
+    #Otherwise, if the key is less than the current key keep on booping left
+    if key < current.key
+      find_helper(current.left, key)
+    #If not, boop right
+    else
+      find_helper(current.right, key)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) -- because we will always need to go throught ALL nodes
+  # Space Complexity: O(n) -- because we will always need to go throught ALL nodes
   def inorder
-    raise NotImplementedError
+    #Return an empty array if our root is nil
+    return [] if @root == nil
+    #Otherwise, call in reinforcements, and tell them where they can put the results
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(node, list)
+    #If where I am is nil return what we already collected
+    return list if node.nil?
+    #Check the window
+    inorder_helper(node.left, list)
+    #If it's not nil, add it to the list
+    list << {key: node.key, value: node.value}
+    #Check the wall
+    inorder_helper(node.right, list)
+  end
+
+  # Time Complexity: O(n) -- because we will always need to go throught ALL nodes
+  # Space Complexity: O(n) -- because we will always need to go throught ALL nodes
   def preorder
-    raise NotImplementedError
+    return [] if @root == nil
+    return preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(node, list)
+    return list if node.nil?
+    #Put your mask on first and THEN check children
+    list << {key: node.key, value: node.value}
+    preorder_helper(node.left, list)
+    preorder_helper(node.right, list)
+
+    return list
+  end
+
+  # Time Complexity: O(n) -- because we will always need to go throught ALL nodes
+  # Space Complexity: O(n) -- because we will always need to go throught ALL nodes
   def postorder
-    raise NotImplementedError
+    return [] if @root == nil
+    return postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(node, list)
+    return list if node.nil?
+    #Check for children BEFORE running out of the burning building
+    postorder_helper(node.left, list)
+    postorder_helper(node.right, list)
+    list << {key: node.key, value: node.value}
+    return list
+  end
+
+  # Time Complexity: O(n) -- because we will always need to go throught ALL nodes
+  # Space Complexity: O(n) -- because we will always need to go throught ALL nodes
   def height
```

Comment:
👍 , but O(h) in space complexity because the call stack will max out at the height of the tree.